apiserver/internal/handlers/resources/download/download.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package download
+
+import (
+	"context"
+	"crypto/sha512"
+	"encoding/hex"
+	"io"
+	"os"
+
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/errors"
+)
+
+// FileSystem defines a file system for modifying files on a users system.
+type FileSystem interface {
+	// CreateTemp creates a new temporary file in the directory dir,
+	// opens the file for reading and writing, and returns the resulting file.
+	CreateTemp(dir, pattern string) (*os.File, error)
+
+	// Open opens the named file for reading.
+	Open(name string) (*os.File, error)
+
+	// Remove removes the named file or (empty) directory.
+	Remove(name string) error
+}
```

Comment:
I use to do this, but in the end, does this actually buy you anything, only because tmp directories exist and you can point your tests at them?

Reply:
For testing I would say it was pretty useful, I caught a few bugs and it makes tests clearer, if you're actually using temp directories all the time theres always a chance your tests are wrong and you're spamming files in there every time you do a unit test.

Reply:
Like the closer not actually catching the close in the error case, I had forgotten to name the return types so the Remove was not being called. Sure, I could have down `os.Stat` to check if its gone or not, but it makes it a load cleaner to just see that the mock wasn't called.